pkg/controller/noderef/noderef_controller.go

```diff
@@ -0,0 +1,183 @@
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+var _ reconcile.Reconciler = &ReconcileNodeRef{}
+
+// ReconcileNodeRef reconciles a Machine object.
+type ReconcileNodeRef struct {
+	client.Client
+	scheme *runtime.Scheme
+}
+
+// Reconcile watches Machines.
+func (r *ReconcileNodeRef) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	// Fetch the NodeRef instance
+	machine := &clusterv1alpha1.Machine{}
+	err := r.Get(context.TODO(), request.NamespacedName, machine)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	// Check that the Machine hasn't been deleted or in the process.
+	if !machine.DeletionTimestamp.IsZero() {
+		return reconcile.Result{}, nil
+	}
+
+	// Check that the Machine has a cluster label.
+	if machine.Labels[clusterv1alpha1.MachineClusterLabelName] == "" {
+		klog.Warningf("Machine %q in namespace %q doesn't specify %q label, cannot set nodeRef", machine.Name, machine.Namespace, clusterv1alpha1.MachineClusterLabelName)
+		return reconcile.Result{}, nil
+	}
+
+	// Check that the Machine has a valid ProviderID.
+	if machine.Spec.ProviderID == nil || *machine.Spec.ProviderID == "" {
+		klog.Warningf("Machine %q in namespace %q doesn't have a valid ProviderID, cannot set nodeRef", machine.Name, machine.Namespace)
+		return reconcile.Result{}, nil
+	}
+
+	cluster, err := r.getCluster(context.TODO(), machine)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	clusterClient, err := remote.NewClusterClient(r.Client, cluster)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	corev1Client, err := clusterClient.CoreV1()
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Get the Node reference.
+	nodeRef, err := r.getNodeReference(corev1Client, machine)
+	if err != nil {
+		if err == ErrNodeNotFound {
+			return reconcile.Result{RequeueAfter: 10 * time.Second}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	// Update Machine.
+	machine.Status.NodeRef = nodeRef
+	if err := r.Client.Update(context.TODO(), machine); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (r *ReconcileNodeRef) getCluster(ctx context.Context, machine *clusterv1alpha1.Machine) (*clusterv1alpha1.Cluster, error) {
+	cluster := &clusterv1alpha1.Cluster{}
+	key := client.ObjectKey{
+		Namespace: machine.Namespace,
+		Name:      machine.Labels[clusterv1alpha1.MachineClusterLabelName],
+	}
+
+	if err := r.Client.Get(ctx, key, cluster); err != nil {
+		return nil, err
+	}
+
+	return cluster, nil
+}
+
+func (r *ReconcileNodeRef) getNodeReference(client corev1.NodesGetter, machine *clusterv1alpha1.Machine) (*apicorev1.ObjectReference, error) {
+	listOpt := metav1.ListOptions{}
+
+	for {
+		nodeList, err := client.Nodes().List(listOpt)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, node := range nodeList.Items {
+			if node.Spec.ProviderID == *machine.Spec.ProviderID {
```

Comment:
Match based on segments of the provider id - format is `$cloudProviderName://$optional/$segments/$etc/$providerID`